lms/djangoapps/courseware/transformers.py

```diff
@@ -0,0 +1,92 @@
+"""
+Grades Transformer
+"""
+from django.contrib.auth.models import AnonymousUser
+from django.test.client import RequestFactory
+
+from openedx.core.lib.block_structure.transformer import BlockStructureTransformer
+from . import module_render
+
+
+class GradesBlockTransformer(BlockStructureTransformer):
+    """
+    The GradesBlockTransformer collects grading information and stores it on
+    the block structure.
+
+    No runtime transformations are performed.
+
+    The following values are stored as xblock_fields on their respective blocks in the
+    block structure:
+
+        due: (datetime) when the problem is due.
+        graded: (boolean)
+        has_score: (boolean)
+        weight: (numeric)
+
+    Additionally, the following value is calculated and stored as a transformer_block_field
+    for each block:
+
+        max_score: (numeric)
+    """
+    VERSION = 1
+    FIELDS_TO_COLLECT = [u'due', u'graded', u'has_score', u'weight']
+
+    @classmethod
+    def name(cls):
+        """
+        Unique identifier for the transformer's class;
+        same identifier used in setup.py.
+        """
+        return u'grades'
+
+    @classmethod
+    def collect(cls, block_structure):
+        """
+        Collects any information that's necessary to execute this
+        transformer's transform method.
+        """
+        block_structure.request_xblock_fields(*cls.FIELDS_TO_COLLECT)
+        cls._collect_max_scores(block_structure)
+
+    def transform(self, block_structure, usage_context):
+        """
+        Perform no transformations.
+        """
+        pass
+
+    @classmethod
+    def _collect_max_scores(cls, block_structure):
+        """
+        Collect the `max_score` for every block in the provided `block_structure`.
+        """
+        for module in cls._iter_xmodules(block_structure):
+            cls._collect_max_score(block_structure, module)
+
+    @classmethod
+    def _collect_max_score(cls, block_structure, module):
+        """
+        Collect the `max_score` from the given module, storing it as a
+        `transformer_block_field` associated with the `GradesBlockTransformer`.
+        """
+        score = module.max_score()
+        block_structure.set_transformer_block_field(module.location, cls, 'max_score', score)
+
+    @staticmethod
+    def _iter_xmodules(block_structure):
+        """
+        Loop through all the blocks locators in the block structure, and retrieve
+        the module (XModule or XBlock) associated with that locator.
+
+        For implementation reasons, we need to pull the max_score from the
+        XModule, even though the data is not user specific.  Here we bind the
+        data to an AnonymousUser.
+        """
+        request = RequestFactory().get('/dummy-collect-max-grades')
+        request.user = AnonymousUser()
+        request.user.known = False
```

Comment:
Please add a comment here of why we need to do set `known`.
Also, @ormsbee or @cpennington can you confirm whether checking for the `known` field on the user is the right thing to happen in [this line in module_render](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/module_render.py#L828)?  It is set by `get_course_lti_endpoints` and `handle_xblock_callback_noauth`.  We are using that same mechanism here to avoid user-specific checking in the Collect phase.